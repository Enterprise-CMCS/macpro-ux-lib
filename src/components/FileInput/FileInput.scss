@use "../../assets/theme/macbis-colors" as c;

// .usa-form-group {
//   .dropzone {
//     border: black 1px dashed;
//   }

//   .files-uploaded {
//     background-color: c.$primary-alt-lightest;
//   }

//   .file {
//     border-top: 1px white solid;
//     border-bottom: 1px white solid;
//     height: 40px;
//   }
// }
