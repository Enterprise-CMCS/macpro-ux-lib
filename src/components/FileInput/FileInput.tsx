import React, { ChangeEventHandler, useLayoutEffect, useRef } from "react";
import fileInput from "../../../node_modules/@uswds/uswds/packages/usa-file-input/src";

type IntrinsicElements = JSX.IntrinsicElements["input"];

export interface FileInputProps extends IntrinsicElements {
  id: string;
  name: string;
  multipleFiles?: boolean;
  label?: string;
  acceptedFileTypes?: string[];
  hintText?: string;
  disabled?: boolean;
  error?: boolean;
  errorMessage?: string;
  onChange?: ChangeEventHandler<HTMLInputElement>;
}

/**
 * **FileInput Component**
 *
 * @param {string}    id                  The id to uniquely identify the file input.
 * @param {string}    name                Name of the file input field.
 * @param {boolean}   multipleFiles       Controls if the input accepts multiple files.
 * @param {string}    label               Used to label the file input in the UI.
 * @param {string[]}  acceptedFileTypes   Array of string file types accepeted in this file input.
 * @param {string}    hintText            Text to show to the user as a guide of the type of accepateble files.
 * @param {boolean}   disabled            Controls whether the file input is disabled.
 * @param {boolean}   error               Triggers error message and error styling.
 * @param {string}    errorMessage        Message to show when an error is present.
 * @param {string}    onChange            Function that is called when the File Input changes.
 *
 */

export const FileInput: React.FC<FileInputProps> = ({
  id,
  multipleFiles = false,
  label,
  name,
  error = false,
  errorMessage,
  acceptedFileTypes = [
    ".pdf",
    ".doc",
    ".docx",
    ".xlsx",
    ".jpg",
    ".jpeg",
    ".png",
  ],
  hintText,
  onChange,
  ...rest
}) => {
  const fileInputRef = useRef<HTMLSpanElement>(null);
  useLayoutEffect(() => {
    const fileInputElement = fileInputRef.current;
    if (typeof fileInput.off === "function") fileInput.off(fileInputElement);
    if (typeof fileInput.on === "function") fileInput.on(fileInputElement);

    return () => {
      if (typeof fileInput.off === "function") fileInput.off(fileInputElement);
    };
  });

  return (
    <div className={`usa-form-group${error ? " usa-form-group--error" : ""}`}>
      <label
        className={`usa-label${error ? " usa-label--error" : ""}`}
        htmlFor={id}
      >
        {label}
      </label>
      {(hintText || errorMessage) && (
        <div>
          {hintText && <span className="usa-hint"> {hintText}</span>}
          {errorMessage && (
            <span className="usa-error-message" id="file-input-error-alert">
              {errorMessage}
            </span>
          )}
        </div>
      )}
      <input
        id={id}
        className="usa-file-input"
        type="file"
        name={name}
        multiple={multipleFiles}
        onChange={onChange}
        {...rest}
      />
    </div>
  );
};
