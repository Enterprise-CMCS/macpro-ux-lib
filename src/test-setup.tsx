import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
