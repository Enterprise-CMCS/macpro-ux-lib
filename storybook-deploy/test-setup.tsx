import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
