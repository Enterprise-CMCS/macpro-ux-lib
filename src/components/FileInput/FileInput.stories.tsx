import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { FileInput } from "./FileInput";

export default {
  title: "COMPONENTS/FileInput",
  component: FileInput,
  args: {},
  argTypes: {
    id: {
      description: "The id to uniquely identify the file input.",
    },
    name: {
      description: "Name of the input field.",
    },
    multipleFiles: {
      description: "Controls if the input accepts multiple files.",
    },
    label: {
      description: "Used to label the file input in the UI.",
    },
    acceptedFileTypes: {
      description: "Array of string file types accepeted in this file input.",
    },
    hintText: {
      description:
        "Text to show to the user as a guide of the type of accepateble files.",
    },
    disabled: {
      description: "Controls whether the file input is disabled.",
    },
    error: {
      description: "Triggers error message and error styling.",
    },
    errorMessage: {
      description: "Message to show when an error is present.",
    },
    onChange: {
      description: "Function that is called when the File Input changes.",
      control: false,
    },
  },
  parameters: {
    docs: {
      description: {
        component: "File input allows users to attach one or multiple files.",
      },
    },
  },
} as ComponentMeta<typeof FileInput>;

const Template: ComponentStory<typeof FileInput> = ({ ...rest }) => (
  <FileInput {...rest} />
);

export const Default = Template.bind({});
Default.args = { name: "1", id: "1" };

export const MultipleFiles = Template.bind({});
MultipleFiles.args = {
  hintText: "This input accepts multiple files",
  multipleFiles: true,
  name: "2",
  id: "2",
};

export const ErrorState = Template.bind({});
ErrorState.args = {
  error: true,
  errorMessage: "There is an error uploading your image",
  name: "3",
  id: "3",
};

export const Disabled = Template.bind({});
Disabled.args = { disabled: true, name: "4", id: "4" };
