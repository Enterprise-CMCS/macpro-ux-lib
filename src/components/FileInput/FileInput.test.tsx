import React from "react";
import { fireEvent, screen, render } from "../../test-setup";
import { FileInput } from "./FileInput";

describe("Tests for the FileInput component", () => {
  it("should render", () => {
    render(<FileInput name="test-1" id="1" data-test-id="file-input" />);

    const fileInput = screen.getByRole("file-input");
    expect(fileInput).toBeInTheDocument();
  });

  it("should render", () => {
    render(
      <FileInput
        name="test-4"
        id="4"
        data-test-id="file-input"
        multipleFiles
        hintText="Test"
        label="test label"
      />
    );
    const fileInput = screen.getByRole("file-input");
  });

  it("should render", () => {
    render(
      <FileInput
        name="test-5"
        id="5"
        data-test-id="file-input"
        disabled
        error
        errorMessage="error"
      />
    );
  });
  const fileInput = screen.getByRole("file-input");
});

describe("compontent snapshots", () => {
  it("default search", () => {
    const { container } = render(
      <FileInput
        name="test-2"
        id="2"
        data-test-id="file-input"
        multipleFiles
        label="test"
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("big search", () => {
    const { container } = render(
      <FileInput
        name="test-2"
        hintText="hint-text"
        onChange={() => {}}
        id="2"
        data-test-id="file-input"
      />
    );
    expect(container).toMatchSnapshot();
  });
});
