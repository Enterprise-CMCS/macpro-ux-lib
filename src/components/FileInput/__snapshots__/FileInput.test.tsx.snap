// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compontent snapshots default file input that allows multiple 1`] = `
<div>
  <div
    class="usa-form-group"
  >
    <label
      class="usa-label"
      for="2"
    >
      test
    </label>
    <input
      class="usa-file-input"
      data-test-id="file-input"
      id="2"
      multiple=""
      name="test-2"
      type="file"
    />
  </div>
</div>
`;

exports[`compontent snapshots file input with hint and onchange 1`] = `
<div>
  <div
    class="usa-form-group"
  >
    <label
      class="usa-label"
      for="2"
    />
    <div>
      <span
        class="usa-hint"
      >
         
        hint-text
      </span>
    </div>
    <input
      class="usa-file-input"
      data-test-id="file-input"
      id="2"
      name="test-2"
      type="file"
    />
  </div>
</div>
`;
